### DISCUSSION TITLE
Tests with WebsocketCommunicator and Django ORM never complete

### DISCUSSION THREAD

**Discussion by emammadov-cpr** [django/channels#2045] [Help!]
Not 100% sure if this is a question for channels or django community. Starting here. 

I am working on writing tests that make use of `WebsocketCommunicator` while also making database calls in the same test case. If the database calls are made before connecting via `WebsocketCommunicator` the test completes. But if the database calls are made after the connection has been established, the test never completes. Minimal setup:

The following test case never completes:
```
from django.test import TransactionTestCase

class TestClass(TransactionTestCase):
    async def test_case(self):
        test_client = WebsocketCommunicator(application, "/api/wamp/", subprotocols=['wamp.2.json'])
        await test_client.connect()

        await call_db_create()
```
where 
```
@database_sync_to_async
def call_db_create(**kwargs):
    result = models.Pizza.objects.create(
        name=f"test_pizza_\u263a_{rand_str()}"
    )
    return result
```

With the switched order of operations:
```
class TestClass(TransactionTestCase):
    async def test_case(self):
        await call_db_create()
        
        test_client = WebsocketCommunicator(application, "/api/wamp/", subprotocols=['wamp.2.json'])
        await test_client.connect()
```
the test completes. 

Consumer:
```
class MyConsumer(WebsocketConsumer):
    def connect(self):
        if "wamp.2.json" in self.scope['subprotocols']:
            self.accept(subprotocol="wamp.2.json")
        else:
            self.close()
```

application:
```
application = ProtocolTypeRouter({
    "http": get_asgi_application(),
    "websocket": AuthMiddlewareStack(
        URLRouter(
            [
                re_path(r'api/wamp/$', consumers.MyConsumer.as_asgi()),
             ]
        )
    ),
})
```

Version:
```
channels==3.0.5
asgiref==3.6.0
django==4.2.0
```

Couple other things that I have found out:
- adding any delay after connecting with the test client and before making a db call, resolves the issue
```
        await test_client.connect()
        time.sleep(1)
        await call_db_create()
```
- running the test case with either the db call or the connect call removed, shows no issue.

Searching online I haven't found any description that matches the test setup where the issue happens. Usually the db calls are made before the test client is connected.In which case it works. 

What could be preventing the async test from completing as described in the initial setup?

**Comment by carltongibson** [comment]
Hi @emammadov-cpr — Nice example. 

> What could be preventing the async test from completing as described in the initial setup?
Not 100% sure at first glance. You'll need to dig in and see exactly what's getting blocked where. 

> adding any delay after connecting with the test client and before making a db call, resolves the issue

So is the first task blocking the second, unless it's able to finish cleanly? 🤔

**Reply by emammadov-cpr** [reply]
Hi @carltongibson 
Yes, at least from the logging that is what gather. I can confirm that the first task appears to complete fine because I get the log from the consumer's connect that the client has connected. I didn't show the logging inside the consumer to keep the example simple. But during the test I do get that log. So the first task completes.

**Comment by carltongibson** [comment]
The `close()` call would likely terminate at least the consumer but maybe the event loop in the test... is that close being triggered? 🤔

**Reply by emammadov-cpr** [reply]
Double checked that `close()` is not being triggered on the consumer. And the connections is always accepted. If the event loop was getting terminated, adding the delay between the two tasks wouldn't have helped. 

Could the next place to look may be `SyncToAsync` implementation? Thinking if its worth trying to find where the blocking may be happening.

**Reply by carltongibson** [reply]
I'd put logging into your `call_db_create`... — are you getting that far. (And then each step..., and so on.)
